### ISSUE TITLE
Python package issue in cross compiling environment  [Illegal instruction]

### ISSUE THREAD

**Issue by pramodskumbhar** [LLNL/spack#5112]
Hello,

Suppose I have package which specifies python dependency as `depends_on('python@2.6:', when='+python', type='link')`.

I am building this on cross compiling environment where python is built for compute node:
```
    python:
        paths:
            python@2.7.13 arch=bgq-cnk-ppc64: /mypath/interpreters/python/2.7.13/powerpc64-bgq-linux/
```

The `build` phase finished fine but then I see:

```
==> Error: ProcessError: Command exited with status -4:
    '/mypath/interpreters/python/2.7.13/powerpc64-bgq-linux/bin/python2.7' '-c' 'from distutils.sysconfig import get_config_var; print get_config_var('"'"'prefix'"'"')'
/gpfs/home/kumbhar/spack/var/spack/repos/builtin/packages/python/package.py:410, in get_config_var:
     403      def get_config_var(self, key):
     404          """Returns the value of a single variable. Wrapper around
     405          ``distutils.sysconfig.get_config_var()``."""
     406
     407          cmd = 'from distutils.sysconfig import get_config_var; '
     408          cmd += self.print_string("get_config_var('{0}')".format(key))
     409
  >> 410          return self.command('-c', cmd, output=str).strip()
```

As shown in the stack trace above, `package.py` try to execute python built for compute node resulting in error:

```
 '/mypath/interpreters/python/2.7.13/powerpc64-bgq-linux/bin/python2.7' '-c' 'from distutils.sysconfig import get_config_var; print get_config_var('"'"'prefix'"'"')'
Illegal instruction (core dumped)
```
Oops..can't execute compute node's python here!

**Comment by pramodskumbhar** [comment]
I see same issue even for `type=('build', 'link')` (I don't see why adding `build` dependency will change the behaviour though)

**Comment by adamjstewart** [comment]
The difference is that `build` adds the Python that Spack installs to the front of the `PATH`. I was hoping that would solve it. Not sure how to solve this one. Anybody else?

**Comment by pramodskumbhar** [comment]
if you have any workaround, please let me know and I would be happy to test (have trouble on different machines including alcf).

**Comment by scheibelp** [comment]
I can start looking at this on 8/22 (maybe 8/21)

> But I don't think that it should have to be a build type to work.

According to #3768 at least I assume that if you want to run a dependency at build time it should have the ```build``` deptype.

**Comment by scheibelp** [comment]
To make sure, it sounds like you don't want Spack to use ```python@2.7.13 arch=bgq-cnk-ppc64``` when concretizing. If that is the case, at the moment I think you will have to set the architecture for ```python``` in the spec provided to ```spack install```. Does something like ```spack spec root ^python arch=...``` give you what you are looking for?

**Comment by pramodskumbhar** [comment]
@scheibelp : Actually I `do want` to use arch=bgq-cnk-ppc64. I have simulation software that links to `libpython.so`. In this case Python is compiled for backend architecture and hence I tried to put build type as `link`. But then Python's package.py actually try to `execute` python binary to find path / libraries etc. (may be @adamjstewart can explain it more precisely)

**Comment by scheibelp** [comment]
OK sorry let me try again: you get the compute-node python in your spec as a dependency (which you want) but when building the dependent package, it fails because it tries to execute the compute node python on the front end node.

**Comment by pramodskumbhar** [comment]
yes.

I briefly looked into this: suppose in dependent package I want to get `include` path or `libs` then following method from `python/package.py` gets executed:

```
    def get_config_var(self, key):
        """Returns the value of a single variable. Wrapper around
        ``distutils.sysconfig.get_config_var()``."""

        cmd = 'from distutils.sysconfig import get_config_var; '
        cmd += self.print_string("get_config_var('{0}')".format(key))

        return self.command('-c', cmd, output=str).strip()
.....

    @property
    def libs(self):
        # Spack installs libraries into lib, except on openSUSE where it
        # installs them into lib64. If the user is using an externally
        # installed package, it may be in either lib or lib64, so we need
        # to ask Python where its LIBDIR is.
        libdir = self.get_config_var('LIBDIR')
```

And here we end up executing `python` binary for compute node...

**Comment by scheibelp** [comment]
OK the issue here may be that another approach is required to get the libraries. I think what you want in this case is the compute-node python's libraries. Right now the only way the python package knows how to retrieve the libraries, headers, or home is by actually running the associated python instance. That won't work for getting the compute node's libraries while running on the frontend. Perhaps packages.yaml could be extended so the user could create entries for libs, headers, and home.

Side note @adamjstewart the ```Python.get_config_var``` command doesn't require setting the build deptype.

**Comment by pramodskumbhar** [comment]
> how to retrieve the libraries, headers, or home is by actually running the associated python instance. That won't work for getting the compute node's libraries while running on the frontend. 

yes

> That won't work for getting the compute node's libraries while running on the frontend. Perhaps packages.yaml could be extended so the user could create entries for libs, headers, and home.

Not sure what would be the best (/ convenient) way. If I understand correctly,  the current implementation of Python package run `python binary itself` because Python may be installed in different location (e.g. brew, on OS X ). May be in `cross compiling environment` we "can" assume that the Python is actually installed in the `specified` path and we should search everything in the specified `prefix` (without executing python binary) ?

**Comment by adamjstewart** [comment]
The situation here should be slightly better now. I modified our Python package to _try_ to call `python` directly and ask sysconfig where the relevant paths are, _except_ failure, and fallback on reasonable default values. Of course, those default values may not work on all systems. I'm not sure of any better way to handle cross-compiling with Python, but let me know if you have any ideas.
